tests/unit-tests/gl/test_tessellation_helpers.cpp

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright © 2020 Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 or 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authored by: William Wold <william.wold@canonical.com>
+ */
+
+#include <glm/gtc/matrix_transform.hpp>
+#include <mir/gl/tessellation_helpers.h>
+#include <mir/test/doubles/mock_renderable.h>
+
+using namespace testing;
+
+namespace geom = mir::geometry;
+namespace mt = mir::test;
+namespace mtd = mir::test::doubles;
+namespace mgl = mir::gl;
+
+class Tessellation : public testing::Test
+{
+public:
+    void SetUp() override
+    {
+        ON_CALL(renderable, screen_position())
+            .WillByDefault(Return(rect));
+    }
+
+    struct BoundingBox
+    {
+        float left, right, top, bottom;
+    };
+
+    static auto bounding_box(mgl::Primitive const& primitive) -> BoundingBox
+    {
+        BoundingBox box{
+            std::numeric_limits<float>::max(),
+            std::numeric_limits<float>::min(),
+            std::numeric_limits<float>::max(),
+            std::numeric_limits<float>::min()};
+
+        for (int i = 0; i < primitive.nvertices; i++)
+        {
+            box.left    = std::min(box.left,    primitive.vertices[i].position[0]);
+            box.right   = std::max(box.right,   primitive.vertices[i].position[0]);
+            box.top     = std::min(box.top,     primitive.vertices[i].position[1]);
+            box.bottom  = std::max(box.bottom,  primitive.vertices[i].position[1]);
+        }
+
+        return box;
+    }
+
+    static void expect_tex_coords_1_or_0(mgl::Primitive const& primitive)
+    {
+        for (int i = 0; i < primitive.nvertices; i++)
+        {
+            for (int axis = 0; axis < 2; axis++)
+            {
+                float const tex_coord = primitive.vertices[i].texcoord[axis];
+                EXPECT_THAT(tex_coord == 0.0 || tex_coord == 1.0, Eq(true));
```

Comment:
If this fails, the report will be:
```plain
Value of: tex_coord == 0.0 || tex_coord == 1.0
Expected: is equal to true
  Actual: false (of type bool)
```

It might be helpful to report what `tex_coord` actually is (and axis). Vis:
```c++
        EXPECT_THAT(tex_coord, AnyOf(Eq(1.0), Eq(0.0))) << "..axis=" << 1;
```
```
Value of: tex_coord
Expected: (is equal to 1) or (is equal to 0)
  Actual: 0.5 (of type float)
..axis=1
```

Reply:
Good idea! I noticed that this wasn't ideal, but I didn't think of the fancier gtest matchers.